### ISSUE TITLE
-All creates neverending loop if total results equals 1

### ISSUE THREAD

**Issue by Ez-G-23** [CrowdStrike/psfalcon#48]
I have noticed an issue whereby `Export-FalconConfig `  seems to fail on certain items. I have two CIDs.  On one of them this command gets stuck on Exporting Prevention Policy. Other items export fine, but when using `-verbose` to see what's going on - it seems to be getting stuck in a loop. In the end it stops without error and I have no file saved and it doesn't continue with the other items.

```
Exporting 'PreventionPolicy'...
VERBOSE: [Invoke-Endpoint] GET https://api.us-2.crowdstrike.com/policy/combined/prevention/v1
VERBOSE: [Read-Meta] trace_id: <Obfuscated>-2e1e47980082
VERBOSE: [Read-Meta] pagination_offset: 7, pagination_total: 7
VERBOSE: [Invoke-Endpoint] GET https://api.us-2.crowdstrike.com/policy/combined/prevention/v1
VERBOSE: [Read-Meta] trace_id: <Obfuscated>-9ca64fad61d7
VERBOSE: [Read-Meta] pagination_offset: 1, pagination_total: 1
VERBOSE: [Invoke-Loop] retrieved  results
VERBOSE: [Invoke-Endpoint] GET https://api.us-2.crowdstrike.com/policy/combined/prevention/v1
VERBOSE: [Read-Meta] trace_id: <Obfuscated>-7fb8c29e45ef
VERBOSE: [Read-Meta] pagination_offset: 1, pagination_total: 1
VERBOSE: [Invoke-Loop] retrieved  results
VERBOSE: [Invoke-Endpoint] GET https://api.us-2.crowdstrike.com/policy/combined/prevention/v1
VERBOSE: [Read-Meta] trace_id: <Obfuscated>-245b0392de22
VERBOSE: [Read-Meta] pagination_offset: 1, pagination_total: 1
VERBOSE: [Invoke-Loop] retrieved  results
VERBOSE: [Invoke-Endpoint] GET https://api.us-2.crowdstrike.com/policy/combined/prevention/v1
VERBOSE: [Read-Meta] trace_id: <Obfuscated>-9ace-f080bdf4a704
VERBOSE: [Read-Meta] pagination_offset: 1, pagination_total: 1
VERBOSE: [Invoke-Loop] retrieved  results
VERBOSE: [Invoke-Endpoint] GET https://api.us-2.crowdstrike.com/policy/combined/prevention/v1
VERBOSE: [Read-Meta] trace_id: <Obfuscated>16b9f206c835
VERBOSE: [Read-Meta] pagination_offset: 1, pagination_total: 1
VERBOSE: [Invoke-Loop] retrieved  results
VERBOSE: [Invoke-Endpoint] GET https://api.us-2.crowdstrike.com/policy/combined/prevention/v1
VERBOSE: [Read-Meta] trace_id: <Obfuscated>42f35de8
VERBOSE: [Read-Meta] pagination_offset: 1, pagination_total: 1
VERBOSE: [Invoke-Loop] retrieved  results
VERBOSE: [Invoke-Endpoint] GET https://api.us-2.crowdstrike.com/policy/combined/prevention/v1
VERBOSE: [Read-Meta] trace_id: <Obfuscated>f8a3fa0e8eba
VERBOSE: [Read-Meta] pagination_offset: 1, pagination_total: 1
VERBOSE: [Invoke-Loop] retrieved  results

```
...and so on

In the other CID, it only gets as far as Exporting DeviceControlPolicy be for showing the same sort of behaviour. So I know it's not specific to prevention policies. This has been tested with an API which has every privilege available and it still occurs.

```
Exporting 'HostGroup'...
VERBOSE: [Invoke-Endpoint] GET https://api.crowdstrike.com/devices/combined/host-groups/v1
VERBOSE: [Read-Meta] trace_id: <Obfuscated>-4d7da3323338
VERBOSE: [Read-Meta] pagination_offset: 67, pagination_total: 67
Exporting 'IOAGroup'...
VERBOSE: [Invoke-Endpoint] GET https://api.crowdstrike.com/ioarules/queries/rule-groups/v1
VERBOSE: [Read-Meta] trace_id: <Obfuscated>-8f61a32bfcc8
VERBOSE: [Read-Meta] pagination_offset: 500, pagination_total: 8
VERBOSE: [Invoke-Endpoint] GET https://api.crowdstrike.com/ioarules/entities/rule-groups/v1
VERBOSE: [Read-Meta] trace_id: <Obfuscated>cb315f261a49
Exporting 'FirewallGroup'...
VERBOSE: [Invoke-Endpoint] GET https://api.crowdstrike.com/fwmgr/queries/rule-groups/v1
VERBOSE: [Read-Meta] trace_id: <Obfuscated>09dbf958fe0b
Exporting 'DeviceControlPolicy'...
VERBOSE: [Invoke-Endpoint] GET https://api.crowdstrike.com/policy/combined/device-control/v1
VERBOSE: [Read-Meta] trace_id: <Obfuscated>-8d715ea0a72d
VERBOSE: [Read-Meta] pagination_total: 1
VERBOSE: [Invoke-Loop] retrieved  results
VERBOSE: [Invoke-Endpoint] GET https://api.crowdstrike.com/policy/combined/device-control/v1
VERBOSE: [Read-Meta] trace_id: <Obfuscated>d7e0398ec65b
VERBOSE: [Read-Meta] pagination_total: 1
VERBOSE: [Invoke-Loop] retrieved  results
VERBOSE: [Invoke-Endpoint] GET https://api.crowdstrike.com/policy/combined/device-control/v1
VERBOSE: [Read-Meta] trace_id: <Obfuscated>-9576753d78fc
VERBOSE: [Read-Meta] pagination_total: 1
VERBOSE: [Invoke-Loop] retrieved  results
VERBOSE: [Invoke-Endpoint] GET https://api.crowdstrike.com/policy/combined/device-control/v1
VERBOSE: [Read-Meta] trace_id: <Obfuscated>-b44f883aef0e
VERBOSE: [Read-Meta] pagination_total: 1
VERBOSE: [Invoke-Loop] retrieved  results
VERBOSE: [Invoke-Endpoint] GET https://api.crowdstrike.com/policy/combined/device-control/v1
VERBOSE: [Read-Meta] trace_id: <Obfuscated>-0bda82260eb3
VERBOSE: [Read-Meta] pagination_total: 1
VERBOSE: [Invoke-Loop] retrieved  results
VERBOSE: [Invoke-Endpoint] GET https://api.crowdstrike.com/policy/combined/device-control/v1
VERBOSE: [Read-Meta] trace_id: <Obfuscated>10a7edc46bba
VERBOSE: [Read-Meta] pagination_total: 1
VERBOSE: [Invoke-Loop] retrieved  results
```
..and so on.

**Comment by Ez-G-23** [comment]
Hello!

I did actually consider this when trying to understand what might be happening, seems logical but my findings didn't fit this exactly - although kind of close:

The CID which was getting stuck on the prevention policies has a few windows prevention policies which are assigned to hosts. It also has a couple of mac policies and Linux policies but crucially these are not assigned and active on any hosts. There is always at least 1 policy + default.

On the CID which was getting stuck on device control policy - again, there is one additional policy aside from the default, but neither of these are applied anywhere either...

**Comment by Ez-G-23** [comment]
When I ran `Get-FalconHost -Limit 1 -All` it worked as expected and retrieved the list of all host IDs in the CID and then stopped.

I deleted the module, redownloaded and added it back and I am still seeing the same behaviour unfortunately.  

If there is anything I can do to help you - let me know.

**Comment by bk-cs** [comment]
Would you mind sending me a [DM on Reddit](https://www.reddit.com/user/bk-CS)? This might require some live troubleshooting because I really can't figure out where this is going wrong based on the existing output.

**Comment by bk-cs** [comment]
Through testing with Ez-G-23, we were able to narrow this down and verify that it only happens when using PSFalcon with PowerShell v5.1, and when you have only one configured policy (besides the default). I'm still working to reproduce it, but if you're encountering this bug you can use PowerShell Core (v6 or newer) and you should be able to `Export-FalconConfig` without issue.

**Comment by bk-cs** [comment]
Ok, I think I've figured this out. It seems that PowerShell 5.1 is unable to `().count` [PSCustomObject] when the value is 1.

PowerShell 5.1
```
PS> $Example = [PSCustomObject] @{ name = 'Example Object' }
PS> $Example.count
PS>
```

PowerShell Core (6+)
```
PS> $Example = [PSCustomObject] @{ name = 'Example Object' }
PS> $Example.count
1
PS>
```

This creates a problem as there's no count for the `Invoke-Loop` function (which repeats requests until all results are received) has nothing to compare against the `Meta.pagination.total` response. **This problem will appear anywhere the** `-All` **parameter is used and the total amount of results is "1"**.

I was able to eliminate this problem by changing lines 1,077-1,079 in `Private\Private.ps1` to force PowerShell to use `Measure-Object` to count the results, rather than using the shorthand `().count`. 

Original:
```
        if ($Loop.Request -and (($Loop.Request.count -lt $Loop.Pagination.total) -or $Loop.Pagination.next_page)) {
            for ($i = $Loop.Request.count; ($Loop.Pagination.next_page -or ($i -lt $Loop.Pagination.total));
            $i += $Loop.Request.count) {
```

Modified:
```
        if ($Loop.Request -and ((($Loop.Request | Measure-Object).count -lt $Loop.Pagination.total) -or
        $Loop.Pagination.next_page)) {
            for ($i = ($Loop.Request | Measure-Object).count; ($Loop.Pagination.next_page -or ($i -lt
            $Loop.Pagination.total)); $i += ($Loop.Request | Measure-Object).count) {
```

Can you update lines 1,077-1,079 in `Private\Private.ps1` to match, reload the PSFalcon module and try `Export-FalconConfig` again in PowerShell 5.1 to see if this resolves your problem too?

If this fixes your issue, I'll include it in the next PSFalcon update. Anyone can apply this same fix between version updates to get around the problem until it's part of an official release.

**Comment by Ez-G-23** [comment]
Nice one bk-cs, this worked! :)

**Comment by bk-cs** [comment]
Thanks! I'll leave this issue open until v2.0.9 has been released with the change.

**Comment by bk-cs** [comment]
I ended up re-writing a lot of the back-end of PSFalcon and moved this change to the latest release, skipping over v2.0.9. This issue is fixed in [v2.1.0-alpha](https://github.com/CrowdStrike/psfalcon/releases/tag/v2.1.0-alpha).
